src/api/objects/InMemorySearchProvider.js

```diff
@@ -0,0 +1,290 @@
+/*****************************************************************************
+ * Open MCT, Copyright (c) 2014-2021, United States Government
+ * as represented by the Administrator of the National Aeronautics and Space
+ * Administration. All rights reserved.
+ *
+ * Open MCT is licensed under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * Open MCT includes source code licensed under additional open source
+ * licenses. See the Open Source Licenses file (LICENSES.md) included with
+ * this source code distribution or the Licensing information page available
+ * at runtime from the About dialog for additional information.
+ *****************************************************************************/
+
+import uuid from 'uuid';
+
+class InMemorySearchProvider {
+    /**
+     * A search service which searches through domain objects in
+     * the filetree without using external search implementations.
+     *
+     * @constructor
+     * @param {Object} openmct
+     */
+    constructor(openmct) {
+        /**
+         * Maximum number of concurrent index requests to allow.
+         */
+        this.MAX_CONCURRENT_REQUESTS = 100;
+        /**
+        * If max results is not specified in query, use this as default.
+        */
+        this.DEFAULT_MAX_RESULTS = 100;
+
+        this.openmct = openmct;
+
+        this.indexedIds = {};
+        this.idsToIndex = [];
+        this.pendingIndex = {};
+        this.pendingRequests = 0;
+
+        this.pendingQueries = {};
+        this.onWorkerMessage = this.onWorkerMessage.bind(this);
+        this.onWorkerMessageError = this.onWorkerMessageError.bind(this);
+        this.onerror = this.onWorkerError.bind(this);
+        this.startIndexing = this.startIndexing.bind(this);
+        this.onMutationOfIndexedObject = this.onMutationOfIndexedObject.bind(this);
+
+        openmct.on('start', this.startIndexing);
```

Comment:
Could we please add a corresponding handler for the `destroy` event that cleans up any worker listeners? This event only happens during test runs, but failing to clean up worker message listeners causes memory leaks and test flake.

There is a future PR that I'm working on which does this for other workers in an attempt to improve our test reliability.